pkg/adaptor/rethinkdb/rethinkdb.go

```diff
@@ -66,290 +62,122 @@ func init() {
-	if r.debug {
-		fmt.Printf("sending all documents for %s\n", table)
-	}
-
-	cursor, err := gorethink.Table(table).Run(r.client)
-	if err != nil {
-		return err
-	}
-	defer cursor.Close()
-
-	var doc map[string]interface{}
-	for cursor.Next(&doc) {
-		if stop := r.pipe.Stopped; stop {
-			return nil
-		}
+// Stop the adaptor
+func (r *RethinkDB) Stop() error {
+	log.With("path", r.path).Infoln("adaptor Stopping...")
+	r.pipe.Stop()
 
-		msg := message.MustUseAdaptor("rethinkdb").From(ops.Insert, r.computeNamespace(table), r.prepareDocument(doc))
-		r.pipe.Send(msg)
-	}
+	close(r.doneChannel)
+	r.wg.Wait()
 
-	if err := cursor.Err(); err != nil {
-		return err
+	if c, ok := r.client.(client.Closer); ok {
+		c.Close()
 	}
 
+	log.With("path", r.path).Infoln("adaptor Stopped")
 	return nil
 }
 
-func (r *RethinkDB) sendChanges(table string, ccursor *gorethink.Cursor) error {
-	defer ccursor.Close()
-	if r.debug {
-		fmt.Printf("sending changes for %s\n", table)
-	}
-	var change rethinkDbChangeNotification
-	for ccursor.Next(&change) {
-		if stop := r.pipe.Stopped; stop {
-			return nil
-		}
-
-		if r.debug {
-			fmt.Printf("change: %#v\n", change)
-		}
-
-		var msg message.Msg
-		if change.Error != "" {
-			return errors.New(change.Error)
-		} else if change.OldVal != nil && change.NewVal != nil {
-			msg = message.MustUseAdaptor("rethinkdb").From(ops.Update, r.computeNamespace(table), r.prepareDocument(change.NewVal))
-		} else if change.NewVal != nil {
-			msg = message.MustUseAdaptor("rethinkdb").From(ops.Insert, r.computeNamespace(table), r.prepareDocument(change.NewVal))
-		} else if change.OldVal != nil {
-			msg = message.MustUseAdaptor("rethinkdb").From(ops.Delete, r.computeNamespace(table), r.prepareDocument(change.OldVal))
-		}
-
-		if msg != nil {
-			r.pipe.Send(msg)
-			if r.debug {
-				fmt.Printf("msg: %#v\n", msg)
-			}
-		}
-	}
+// applyOp applies one operation to the database
+func (r *RethinkDB) applyOp(msg message.Msg) (message.Msg, error) {
+	err := client.Write(r.client, r.writer, message.From(msg.OP(), msg.Namespace(), msg.Data()))
 
-	if err := ccursor.Err(); err != nil {
-		return err
+	if err != nil {
+		r.pipe.Err <- adaptor.NewError(adaptor.ERROR, r.path, fmt.Sprintf("write message error (%s)", err), msg.Data)
 	}
-
-	return nil
+	return msg, err
 }
 
-func (r *RethinkDB) computeNamespace(table string) string {
-	return strings.Join([]string{r.database, table}, ".")
+func (r *RethinkDB) tableFilter(table string) bool {
+	return r.tableMatch.MatchString(table)
 }
 
 // prepareDocument moves the `id` field to the `_id` field, which is more
 // commonly used by downstream sinks. A transformer could be used to do the
 // same thing, but because transformers are not run for Delete messages, we
 // must do it here.
-func (r *RethinkDB) prepareDocument(doc map[string]interface{}) map[string]interface{} {
+func prepareDocument(doc map[string]interface{}) map[string]interface{} {
```

Comment:
I kind of hate that we do this because it makes it quite irritating for someone who is wanting do to a rethinkdb -> rethinkdb pipeline.